### PR TITLE
Use Scope Generator in (almost) all scope loops

### LINKED ISSUE #91: completion of OEnumLiterals (from packages) does not work

### PR DESCRIPTION
**Pull request by Nik-Sch**
added iteration through use clauses in scope generator

fixes #91

### CODE REVIEW

File: test/test_files/test_error_expected/04_generic_package/generic_pkg.vhd [hunk 1]
```diff
@@ -1,9 +1,9 @@
-package generic_pkg is
+package generic_pkg_error is
   generic (
     generic_parameter : integer := 0
     );
 end package;
-package body generic_pkg is
+package body generic_pkg_error is
     constant generic_parameter : integer := 0; --multiple declaration
```
Comment (g0t00) [inline test/test_files/test_error_expected/04_generic_package/generic_pkg.vhd:7]: Maybe rename th e file to generic_pkg_error than?

File: test/test_files/test_no_error/04_generic_package/test_entity_generic.vhd [hunk 2]
```diff
@@ -5,8 +5,8 @@ generic (
 end entity;
 architecture arch of test_entity_generic is
   use test_pkg.all;
-  signal a : integer := generic_parameter;
+  signal a : t_testData;
```
Comment (g0t00) [inline test/test_files/test_no_error/04_generic_package/test_entity_generic.vhd:9]: why did you remove the test for the generic parameter?

File: lib/parser/objects.ts [hunk 3]
```diff
@@ -1068,6 +1068,11 @@ export function* scope(startObject: ObjectBase) {
     if (current instanceof OPackageBody && current.correspondingPackage) {
       yield current.correspondingPackage;
     }
+    if (implementsIHasUseClause(current)) {
+      for (const packages of current.packageDefinitions) {
+        yield packages;
+      }
+    }
     if (current.parent instanceof OFile) {
```
Comment (g0t00) [inline lib/parser/objects.ts:1076]: I really like this change. But this is manually done at other points in the code, doing only this probably creates duplicated warnings/errors or?

Comment (Nik-Sch) [reply lib/parser/objects.ts:1076]: I haven't tested for duplicate messages, you are probably right.
We could change this PR to use the scope generator everywhere? Or should we create another PR which migrates everything to the scope generator and postpone the merge of this PR to after the scope migration PR?

**Comment by Nik-Sch** [comment]
As per discussion, I escalated this PR to replace (almost) all the iterations with the scope generator. This should improve reliability.

**Comment by Nik-Sch** [comment]
[test/test_files/test_error_expected/01/test_pkg_reference.vhd](https://github.com/vhdl-linter/vhdl-linter/pull/107/commits/162eb7176a5502cfe69eabb04f7b61a7b21e4f2b#diff-95738122125db5be285e97f78f80d89992c91596db6a43cacb0089af5e6aeba3) has been removed because this could be read at this location if it were `pkg_test'instance_name` for example and we cannot detect the difference at the moment

**Comment by g0t00** [comment]
> [test/test_files/test_error_expected/01/test_pkg_reference.vhd](https://github.com/vhdl-linter/vhdl-linter/pull/107/commits/162eb7176a5502cfe69eabb04f7b61a7b21e4f2b#diff-95738122125db5be285e97f78f80d89992c91596db6a43cacb0089af5e6aeba3) has been removed because this could be read at this location if it were `pkg_test'instance_name` for example and we cannot detect the difference at the moment

Are you sure it does not need to be `work.pkg_test'instance_name` then?

File: lib/parser/objects.ts [hunk 4]
```diff
@@ -135,6 +135,7 @@ export class OIRange implements Range {
 
 export class ObjectBase {
   lexerToken?: OLexerToken;
+  readonly sourceFile: OFile;
```
Comment (g0t00) [inline lib/parser/objects.ts:138]: rootFile? sourceFile seems weird

**Comment by Nik-Sch** [comment]
> > [test/test_files/test_error_expected/01/test_pkg_reference.vhd](https://github.com/vhdl-linter/vhdl-linter/pull/107/commits/162eb7176a5502cfe69eabb04f7b61a7b21e4f2b#diff-95738122125db5be285e97f78f80d89992c91596db6a43cacb0089af5e6aeba3) has been removed because this could be read at this location if it were `pkg_test'instance_name` for example and we cannot detect the difference at the moment
> 
> Are you sure it does not need to be `work.pkg_test'instance_name` then?

Ok, you are correct, The visibility of package identifier is weird as it is visible within the package itself but not from the outside. I introduced a new scope generator which adds information about the object being directly visible...

File: lib/parser/objects.ts (hunk 4 above)
Comment (Nik-Sch) [reply lib/parser/objects.ts]: yes, makes more sense. Changed that.
Was just going for a differentiation from `rootElement`

File: lib/parser/objects.ts [hunk 5]
```diff
@@ -1026,18 +1034,27 @@ export class OConfiguration extends ObjectBase implements IHasLibraries {
   libraries: OLibrary[] = [];
 }
 export function* scope(startObject: ObjectBase) {
+  for (const iterator of scopeWithVisibility(startObject)) {
+    yield iterator.obj;
+  }
+}
+export function* scopeWithVisibility(startObject: ObjectBase) {
   let current = startObject;
+  let directlyVisable = true;
```
Comment (g0t00) [inline lib/parser/objects.ts:1047]: directlyVisible?

File: lib/parser/objects.ts [hunk 6]
```diff
@@ -1026,18 +1034,27 @@ export class OConfiguration extends ObjectBase implements IHasLibraries {
   libraries: OLibrary[] = [];
 }
 export function* scope(startObject: ObjectBase) {
+  for (const iterator of scopeWithVisibility(startObject)) {
+    yield iterator.obj;
+  }
+}
+export function* scopeWithVisibility(startObject: ObjectBase) {
```
Comment (g0t00) [inline lib/parser/objects.ts:1045]: What is happening here? I am confused? Can you add a comment?

The one gives visibility information, the other strips that info away?
Can you add comment at two both functions at least?
And Can we keep it as a single iterator and yield [obj, directlyVisible] so that you can easily desctrucure it? (as array not object?)

Comment (g0t00) [reply lib/parser/objects.ts]: I took the liberty to cleanup a bit. Feel free to close resolve and merge If you think it is fine.
